app/javascript/react/textual_summary/icon_or_image.jsx

```diff
@@ -0,0 +1,42 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+
+/**
+ * Render icon or an image with a title
+ */
+export default class IconOrImage extends React.Component {
+  constructor(props) {
+    super(props);
+    //console.log(props);
+  }
+
+  renderIcon() {
+    // %i{:class => item[:icon], :title => item[:title]}
+    return <i className={this.props.icon} title={this.props.title} />;
+  }
+
+  renderImage() {
+    // = encodable_image_tag(item[:image], :alt => item[:title], :title => item[:title])
+    return <img src={this.props.image} alt={this.props.title} title={this.props.title} />;
+  }
+
+  render() {
+    if (this.props.image == null) {
+      return this.props.icon == null ? '' : this.renderIcon();
+    }
+    return this.renderImage();
+  }
+}
+
+IconOrImage.propTypes = {
+  title: PropTypes.string,
+  image: PropTypes.string,
+  icon: PropTypes.string,
+};
+
+IconOrImage.defaultProps = {
+  title: null,
+  image: null,
+  icon: null,
+};
+
```

Comment:
can be written as:
```js
import * as React from 'react';
import PropTypes from 'prop-types';

/**
 * Render icon or an image with a title
 */
const IconOrImage = ({ icon, title, image }) => {
  const renderIcon = () => <i className={icon} />;

  const renderImage = () => <img src={image} alt={title} title={title} />;

  return image ? renderIcon() : renderImage();
};

IconOrImage.propTypes = {
  title: PropTypes.string,
  image: PropTypes.string,
  icon: PropTypes.string,
};

IconOrImage.defaultProps = {
  title: null,
  image: null,
  icon: null,
};

export default IconOrImage;
```

Reply:
yes, I will move the methods inside the class (I did that in some places already).

But the line `return image ? renderIcon() : renderImage();` is not the same as my original code. This simplification would not work.

Reply:
sorry i did not test the code, what did not work?  how about
```js
return (image === null ?  renderIcon() : renderImage());
```

Reply:
no ;-)

in case of both not present I need an empty string. And nested ternary is discouraged so I have that `if`